### PR TITLE
Implement code signing and basic pipeline to create draft releases.

### PR DESCRIPTION
**Pull request by gdasher**
# Implements code signing and basic pipeline to create draft releases #

## 🗣 Description ##

* Adds a github action that will create draft releases artifacts in zip format with signed powershell code
* Can be extended in future with other steps (testing, packaging, etc) and refined as appropriate
* Uses draft releases so that release notes can be manually refined as necessary.

## 💭 Motivation and context ##

This closes issue #95.  It enables us to release signed code that will run on windows without unblock-file. 

## 🧪 Testing ##

Initial testing was completed in a forked branch with an untrusted CA.   PR has been tested by running action from dev branch using real certificate.  Release artifacts run successfully without unblock-file in an environment that enforces signing for MOTW scripts.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

### CODE REVIEW

**Comment by gdasher** [comment]
Here is a sample draft release created by the pipeline: https://github.com/cisagov/ScubaGear/releases/tag/untagged-a118b16a6b44a63dcb82

In my (limited) testing on LAN-A, I confirmed that the code was able to be run without running unblock-file, which had not previously been the case with the default powershell execution policy.

File: .github/workflows/run_release.yaml
Comment (gdasher) [inline .github/workflows/run_release.yaml:28]: I created an extra level of directory structure to be doubly sure that the private key doesn't end up in the release artifact (even if the Remove-Item failed, say).

Before this is committed, I will look at eliminating that from the zip file when its created, but flagging for now as FYI.

Comment (gdasher) [reply .github/workflows/run_release.yaml:28]: This is done now.

**Comment by nanda-katikaneni** [comment]
Attempted to use the tool from zipped/downloaded branch code on a Windows VM - still have to run the unblock-file before being able to run the ScubaGear.

**Comment by gdasher** [comment]
Interesting, that wasn't my experience.  Ping me on the slack when you have a few minutes and we can hopefully figure out what is going on.

**Comment by ethanb-cisa** [comment]
I checked `Orchestrator.psm1` on my personal desktop and it validates. 

@gdasher should we sign the `ScubaGear.psd1` manifest file as well?

**Comment by gdasher** [comment]
I did notice that on my work laptop it doesn't run without prompting for approval.  I think the issue is because the PFX file ethan created doesn't have the intermediate that signed our CA: DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1

Work computer only trusts CN = DigiCert Trusted Root G4

I will chase down the interemdiate and give that to @ethanb-cisa to upload as a new secret.

On the manifest, I'm not sure.  Let me research it.

**Comment by nanda-katikaneni** [comment]
> 

@schrolla Interesting - I used the branch release and verified that cert blob at the end of setup.ps1 is there. When I try to execute the Setup.ps1 - getting the same old unblock prompt.

**Comment by gdasher** [comment]
I believe this is implemented correctly and I understand behavior.  

When code is unsigned, you get this error:

Screenshot (10)

When code is signed with an invalid CA, you get this error:

Screenshot (7)

When code is signed with a valid CA and there is a path to the root, you get this prompt:

Screenshot (9)

I think this is similar to the "Publisher" prompt where you consent to the publisher of an app when you first run it.  My experience was that if you consent to the publisher, it persists and you can run code from that publisher (CISA in this case) going forward.  So I think this is actually correct behavior.

If others can confirm this behavior, namely that after you "approve" the publisher, it runs as expected without requiring unblock  file, even if the signing policy is RemoteSigned or AllSigned, I think we should be good.

I also changed the code to use Set-AuthenticodeSignature rather than signtool.exe because I think it is slightly clearly code.

**Review by nanda-katikaneni — APPROVED** [review]
Did additional tests with updated test2 release: tested on VM's with PowerShell 5.1 with different execution policies. With execution policy set to "Remote Signed and AllSigned" - seeing the correct behavior (i.e. once the remote publisher is approved, user is able to execute the tool without having to unblock). With "Unrestricted" execution policy, user is prompted to unblock for each use. Maybe we should update the README to note this behavior with unrestricted execution policy (can be done through a separate issue).

No other issues/comments - approving the pr.

File: .github/workflows/run_release.yaml
Comment (gdasher) [reply .github/workflows/run_release.yaml:50]: Yeah, I was thinking the same actually.  Done.

File: .github/workflows/run_release.yaml
Comment (gdasher) [reply .github/workflows/run_release.yaml]: 1) Not to my knowledge.  The only gov time server I know is NIST's, but its an NTP server and doesn't speak this protocol.  So I just picked one of the vendor time servers that is publicly available and seemed widely used.

2) Good point and I agree.  Fixed.

3) Done.  I saw no evidence of throttling.

4) Yes, I agree.  I had actually done that previously with signtool but lost it with the changed.  Fixed.

**Comment by gdasher** [comment]
Addressed all comments.  test3 release draft has code implementing the changes.  I will test tomorrow to verify, but don't expect any issues.

**Pull request merged**
